minLength & maxLength + ES6 Syntax

nellfs opened the pull request:
* minLength and maxLength

* es6 syntax

* syntax highlighting on README.md

<!-- Please don't delete this template -->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

## 1.4.0 - minLength And maxLength & ES6 Syntax

*Added 2 new options `minLength` and `maxLength`. Updates to modern Javascript Syntax. "Closes #13"*

## What are the specific steps to test this change?

> 1. Clone the repository 
> 2. Run ```npm run install``` and then ```npm run test``` *or create a link and use it like a normal javascript library*)

## What kind of change does this PR introduce?

- [ ] Bug fix
- [x] New feature
- [x] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [x] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [x] The changelog is updated
- [x] Related documentation has been updated
- [x] Related tests have been updated

BoDonkey reviewed (changes requested):
Overall, the refactoring is great and the new option is almost there! One thing I pointed out in the issue notes is that any potential errors created by the `minLength` option being out of range need to be handled in some way. Right now, submitting a non-string value for the `join` option results in that option being replaced with `" "`. This is a possibility for someone entering a number that is larger than the largest word in the dictionary - say `minLength: 74`. Right now, if you were to try that you would create an infinite loop. This needs to be addressed in some way.

nellfs commented:
I made the requested changes and the treatment of possible errors regarding `minLength` and `maxLength`, if they receive a value other than a number they must return the minimum possible value and the maximum possible value for a word. The same happens if their value passes the maximum or minimum, a loop should not happen anymore since the values ​​are `limited` before being used.

BoDonkey commented on index.js:
```diff
@@ -277,19 +32,38 @@ function words(options) {
   }
 
+  // Generate one word with limits
+  if (
```
In this code block, only the first if statement is hit. The two blocks are essentially the same. In fact, in the current tests, neither of these blocks get hit because there are always additional options being passed. Can you eliminate the first code block, please.

BoDonkey commented on tests/test.js:
```diff
@@ -1,131 +1,178 @@
-    });
+    // with these seeds, the number of words should 5 and 3
+    assert.ok(words.length != words2.length, "number of words is different");
+  });
```
Can you please write 4 additional tests. One is to test that passing just `minLength` works, one showing that `maxLength` alone works. This tests your formally redundant code blocks. One showing that passing a `maxLength` larger than the biggest word does not result in an endless loop, and finally, one that shows passing a `minLength` larger than the longest word doesn't result in an endless loop.

BoDonkey reviewed (changes requested):
This looks really good! I still have a few code changes and a question before I accept.
The requested code changes are in comments below.
I have a question about the typescript file. Since you changed how the dictionary is exported, do you also have to change the declarations file? I'm not very TypeScript fluent, so the answer here could be "no".
So, you should not open a new PR, just push changes to your existing repo that you want to merge. Then let me know on Discord or twitter when you want a re-review if you can't see the reviewers section.

nellfs commented:
> Since you changed how the dictionary is exported, do you also have to change the declarations file? 

The necessary changes have already been made.

>Can you add an example, with comment, about what happens if you use a maxLength that is larger than any word in the dictionary, please.

Done. I added a little comment on some others too.

>In this code block, only the first if statement is hit. The two blocks are essentially the same. In fact, in the current tests, neither of these blocks get hit because there are always additional options being passed. Can you eliminate the first code block, please.

Sorry, I just didn't realize I left two different ones. I changed the block, they were no longer needed as they were originally thought, it's different now.

>Can you please write 4 additional tests. One is to test that passing just minLength works, one showing that maxLength alone works. This tests your formally redundant code blocks. 

Actually these 2 were already included.

> One showing that passing a maxLength larger than the biggest word does not result in an endless loop, and finally, one that shows passing a minLength larger than the longest word doesn't result in an endless loop.

They were included now.

* In addition, I made a small change to index.d.ts, I noticed that `count` was mandatory, now it's optional since you don't always want to specify a `count`

BoDonkey commented:
> > Can you please write 4 additional tests. One is to test that passing just minLength works, one showing that maxLength alone works. This tests your formally redundant code blocks.
> 
> Actually these 2 were already included.

So.... kind of. If I remember the double logic block that you had before, the test would come out differently depending on whether the `exactly` option was passed. The existing tests were looking at `exactly` and `minLength` or `maxLength`. While I like to see testing 'in isolation', so only one option passed for testing (are you testing `exactly`, `minLength`, or both?), I think this is fine as is.

boutell commented on index.js:
```diff
@@ -277,70 +41,76 @@ function words(options) {
-module.exports = words;
 // Export the word list as it is often useful
-words.wordList = wordList;
+export default words;
```
The word list is no longer exported. Although I realize we are moving to ESM here already and it must already be a new major version, which allows backward compatibility breaks, we prefer that any exports be made from the root file so that developers are not told to directly import sub-files, which can lead to importing of files that are not meant to be part of a public and maintained API.

So, a second, non-default export for the wordList is called for at this level.

nellfs replied:
Isolated tests for `minLength` and `maxLength` have been added. I also added an `AssertionError message` in the tests because I had forgotten about it. 

In the end I thought it best to move wordList[] back to index, I could do a different export, something like a barrel export or something else, but I thought it best to keep things as they are and leave everything centered in `index.js`.

nellfs commented:
> The existing tests were looking at `exactly` and `minLength` or `maxLength`. While I like to see testing 'in isolation', so only one option passed for testing (are you testing `exactly`, `minLength`, or both?), I think this is fine as is.

Isolated tests for `minLength` and `maxLength` have been added. I also added an `AssertionError message` in the tests because I had forgotten about it. 😅

BoDonkey reviewed (approved):
Approved
